teps/0125-add-credential-filter-to-entrypoint-logger.md

```diff
@@ -0,0 +1,337 @@
+---
+status: proposed
+title: Add credential filter to entrypoint logger
+creation-date: '2022-10-27'
+last-updated: '2022-10-27'
+authors:
+- '@useurmind'
+---
+
+# TEP-0125: Add credential filter to entrypoint logger
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Currently tekton pipeline steps are printing out any secrets in plaintext. This 
+can be intended output from shell scripts but also accidental output from e.g. 
+the `set -x` shell command.
+
+Both is a security issue as logs are often propagated to log aggregators from a 
+central daemon. This can leak secrets.
+
+Therefore the tekton entrypoint which is running all pipeline step commands should
+filter secrets automatically where possible. 
+
+## Motivation
+
+Automatically avoid leaking secret values/credentials into the output logs of
+tekton pipelines.
+
+### Goals
+
+Secrets are filtered correctly from the output log in a way that the developer 
+of the pipeline does not need to configure this.
+
+Especially output from the `set -x` command should also be filtered correctly.
+
+### Non-Goals
+
+We only target secret values that tekton can automatically redact. We do not want 
```

Comment:
(non-blocking - just another idea) You may also be able to add additional rules for certain types of well-known secret types (i.e. JWT tokens, GitHub tokens, etc.) similar to https://github.com/zricethezav/gitleaks/blob/master/config/gitleaks.toml

Reply:
I would say this is also a non goal of this TEP. ;)

The tool you linked is meant to detect more like accidentially added secrets in git repositories. That can happen quite easily when you worked with something that puts secrets in the working directory and commit everything to git. You can detect those with pattern matching.

When you run pipelines in tekton you provision secrets on purpose to the pipeline, either by ENV, file or tekton parameter.

You could provide a secret accidentially in a non standard way like putting it directly into an env or parameter. But you would not be able to pattern match this kind of stuff, because different people will use different names for those env vars and parameters.

To not blow this TEP up to much I would very much like to keep the focus on the described functionality. Handling parameters and accidental secrets is a very different topic and more complicated as this will probably also require API changes. The implementation already done and the related PR is open for quite a long time now and I want to finish the whole thing. We are all eager to finally see this working. ;)

Reply:
Thanks @Useurmind, I think stating this as a non goal or future work makes a lot of sense. I appreciate your responsiveness to comments on this TEP and your PR, and I'm sorry your PR has been open for so long-- not the contributor experience we aim for. I'll take another look at it this week.

Reply:
I am sad to say this but I will not work on this feature set or its specification atm. because I do not have time for it.
Feel free to extend the TEP regarding these topics. I will continue working on the original pull request and adapt it to use the downward API instead of an environment variable.

It there aren't any more design problems with the solution I would regard this pull request as done.

Reply:
SGTM - no expectation for you to do any additional implementation work! 🙇 

I'm moreso interested in making sure we leave the door open to add this in the future. +1 to @lbernick's suggestion for making a note of this in a `Future Work`section.

Reply:
I have added the section for future work